[TypeChecker/NameLookup] SE-0463: A few fixes for `SendableCompletionHandlers` feature

xedin opened the pull request:
-  Allow `@objc @implementation` to mismatch on `@Sendable` in parameter positions

   Warn about missing `@Sendable` annotations (even in Swift 6 language mode) to give users and developers time to adopt Swift concurrency.

- Shadowing: strip `@Sendable` annotations for declarations ObjC/@objc declarations to avoid ambiguities until shadowing declarations adopt `@Sendable`.

- Strip Sendable while processing AnyObject lookup results

-  Allow associated type inference to skip `@Sendable` on ObjC witnesses

- Fix `@_dynamicReplacement` to skip sendable annotations on ObjC declarations

Resolves: rdar://85569247

<!--
If this pull request is targeting a release branch, please fill out the
following form:
https://github.com/swiftlang/.github/blob/main/PULL_REQUEST_TEMPLATE/release.md?plain=1

Otherwise, replace this comment with a description of your changes and
rationale. Provide links to external references/discussions if appropriate.
If this pull request resolves any GitHub issues, link them like so:

  Resolves <link to issue>, resolves <link to another issue>.

For more information about linking a pull request to an issue, see:
https://docs.github.com/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue
-->

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[TypeChecker] Allow `@objc @implementation` to mismatch on `@Sendable` in parameter positions

Just like for protocol matching, let's allow `@Sendable` mismatches
in parameter positions to make sure that it's possible for header
authors to add concurrency annotations without breaking clients.

This is especially important for `SendableCompletionHandlers` feature
that makes imported sync completion handler parameters `@Sendable`.

xedin committed:
[NameLookup] Shadowing: strip `@Sendable` annotations for declarations ObjC/@objc declarations

When `SendableCompletionHandlers` feature is enabled, ClangImporter
is going to inject `@Sendable` attribute into the completion handler
parameter's function type. We need to make sure that his new behavior
doesn't break shadowing rules and allow APIs to annotate their ObjC
declarations with `@Sendable` without breaking clients.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test Linux platform

xedin commented on lib/AST/NameLookup.cpp:
```diff
@@ -800,6 +800,77 @@ static CanType removeThrownError(Type type) {
+            [](TypeBase *type) -> std::optional<Type> {
+              if (auto *funcType = type->getAs<AnyFunctionType>()) {
+                auto extInfo = funcType->getExtInfo();
+                if (!extInfo.isSendable())
```
Hm, this is actually tricky because ObjC side could have a `@nonSendable` attribute but we cannot know that here...

beccadax commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -3729,8 +3760,10 @@ class ObjCImplementationChecker {
 
-    if (!matchTypes(getMemberType(req), getMemberType(cand), cand))
-      return MatchOutcome::WrongType;
+    auto matchTypesOutcome =
+        matchTypes(getMemberType(req), getMemberType(cand), cand);
+    if (matchTypesOutcome < MatchOutcome::Match)
+      return matchTypesOutcome;
```
Oh, this is a problem. `matchesImpl()` needs to test for, or at least return, the cases in the order they're declared in by the `MatchOutcome` enum. The way you have it here, you could end up returning `MatchOutcome::WrongSendability` before we check for some of the more severe issues, like `WrongForeignErrorConvention`, so if a declaration has *both* a sendability mismatch *and* one of these errors, we'll end up ignoring the more severe error and emitting the sendability warning instead.

(This is mentioned in the doc comment on `MatchOutcome`, but a comment near the top of `matchesImpl()` too wouldn't be a bad idea.)

One way to resolve that problem would be to store off the outcome for later:
```suggestion
   auto matchTypesOutcome =
        matchTypes(getMemberType(req), getMemberType(cand), cand);
    if (matchTypesOutcome == MatchOutcome::WrongType)
      return MatchOutcome::WrongType;
      
    // ...checks for WrongWritability, WrongRequiredAttr, WrongForeignErrorConvention...

    if (matchTypesOutcome == MatchOutcome::WrongSendability)
      return MatchOutcome::WrongSendability;

    ASSERT(matchTypesOutcome == MatchOutcome::Match && "unexpected matchTypes() return");
```

Another possibility would be to go back to a `bool`-returning `matchTypes()` (the logic with the tri-state return is becoming a little tortured), but add `TypeMatchOptions` parameters where needed, so you can do a sendable-insensitive test for `WrongType` and then a sendable-sensitive one for `WrongSendability`:

```suggestion
    if (!matchTypes(getMemberType(req), getMemberType(cand), cand, 
                    TypeMatchFlags::IgnoreFunctionSendability))
       return MatchOutcome::WrongType;

    // ...checks for WrongWritability, WrongRequiredAttr, WrongForeignErrorConvention...

    if (!matchTypes(getMemberType(req), getMemberType(cand), cand, {}))
       return MatchOutcome::WrongSendability;
```

xedin replied:
✅ + added a test.

beccadax commented on lib/AST/NameLookup.cpp:
```diff
@@ -826,6 +826,15 @@ static void recordShadowedDeclsAfterSignatureMatch(
+    // Strip `@Sendable` annotations for declarations that come from
+    // or are exposed to Objective-C to make it possible to introduce
+    // new annotations without breaking shadowing rules.
+    if (decl->hasClangNode() || decl->getAttrs().hasAttribute<ObjCAttr>()) {
```
IIRC, you're using `hasAttribute<ObjCAttr>()` here, rather than `isObjC()`, because you're trying to keep this narrowly focused on specific backwards-incompatible cases? If so, I would mention that in the comment, because in isolation this condition looks pretty weird.

xedin replied:
Sure!

xedin replied:
✅

beccadax commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -3436,6 +3436,8 @@ class ObjCImplementationChecker {
     WrongRequiredAttr,
     WrongForeignErrorConvention,
 
+    WrongSendability,
```
Nit: I would remove the blank line above this case.

xedin replied:
✅

beccadax commented on lib/Sema/TypeCheckDeclObjC.cpp:
```diff
@@ -3621,70 +3623,99 @@ class ObjCImplementationChecker {
+        return MatchOutcome::WrongSendability;
+    }
+
+    if (matchOpts.contains(TypeMatchFlags::AllowNonOptionalForIUOParam)) {
```
Why add this flag? If your goal is to ensure that we only apply this logic when we have an underlying parameter to check for IUO-ness, surely a null check on `implDecl` would do that without having to manage an extra flag.

xedin replied:
Sure, I can do it that way.

xedin replied:
✅

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility release

xedin commented:
@swift-ci please test source compatibility release

xedin commented:
@swift-ci please test source compatibility

xedin committed:
[ConstraintSystem] Strip Sendable while processing AnyObject lookup results

All of the declarations returned by dynamic `AnyObject` lookup
are `@objc` and hence `@preconcurrency` which means that it should
be possible to introduce `@Sendable` and `any Sendable` annotations
without affecting lookup and shadowing behavior.

xedin committed:
[Sema] Allow associated type inference to skip `@Sendable` on ObjC witnesses

Witness checking itself is allowed to skip `@Sendable` mismatch in this
situation. This is a narrow fix for `SendableCompletionHandlers` feature
to make sure that the behavior doesn't change for non-ObjC witnesses

xedin committed:
[TypeChecker] Fix `@_dynamicReplacement` to skip sendable annotations on ObjC declarations

Refactors `@_dynamicReplacement` attribute verification to consider
only exact matches first, and if there are none, attempt to strip
away sendability on ObjC declarations to make sure that any new
`@Sendable` or `any Sendable` introduced to `@precocurrency` declarations
don't break the overload selection.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test
